### ISSUE TITLE
Plugin v1.12.0 incompatible with SonarQube v9.7.0

### ISSUE THREAD

**Issue by stevehipwell** [mc1arke/sonarqube-community-branch-plugin#674]
**Describe the bug**
The v1.12.0 version of this plugin doesn't seem to be compatible with SonarQube v9.7.0.

**To Reproduce**
Install v1.12.0 to SonarQube v9.7.0.

**Expected behavior**
I'd expect the plugin to load as it does for v9.6.1.

**Screenshots**
> 2022.10.18 09:01:53 INFO  app[][o.s.a.ProcessLauncherImpl] Launch process[WEB_SERVER] from [/opt/sonarqube]: /usr/lib/jvm/java-11-openjdk/bin/java -Djava.awt.headless=true -Dfile.encoding=UTF-8 -Djava.io.tmpdir=/opt/sonarqube/temp -XX:-OmitStackTraceInFastThrow --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.rmi/sun.rmi.transport=ALL-UNNAMED --add-exports=java.base/jdk.internal.ref=ALL-UNNAMED --add-opens=java.base/java.nio=ALL-UNNAMED --add-opens=java.base/sun.nio.ch=ALL-UNNAMED --add-opens=java.management/sun.management=ALL-UNNAMED --add-opens=jdk.management/com.sun.management.internal=ALL-UNNAMED -Dcom.redhat.fips=false -Xmx512m -Xms128m -XX:+HeapDumpOnOutOfMemoryError -javaagent:./extensions/plugins/sonarqube-community-branch-plugin.jar=web -Dhttp.nonProxyHosts=localhost|127.*|[::1] -cp ./lib/sonar-application-9.7.0.61563.jar:/opt/sonarqube/lib/jdbc/h2/h2-2.1.214.jar org.sonar.server.app.WebServer /opt/sonarqube/temp/sq-process11500539336540629664properties
09:01:53.624 [main] INFO com.github.mc1arke.sonarqube.plugin.CommunityBranchAgent - Loading agent
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:513)
	at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:525)
Caused by: java.lang.ClassNotFoundException: org.sonar.server.almsettings.MultipleAlmFeatureProvider
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:315)
	at com.github.mc1arke.sonarqube.plugin.CommunityBranchAgent.redefineEdition(CommunityBranchAgent.java:93)
	at com.github.mc1arke.sonarqube.plugin.CommunityBranchAgent.premain(CommunityBranchAgent.java:56)
	... 6 more
*** java.lang.instrument ASSERTION FAILED ***: "result" with message agent load/premain call failed at src/java.instrument/share/native/libinstrument/JPLISAgent.c line: 422
FATAL ERROR in native method: processing of -javaagent failed, processJavaStart failed
2022.10.18 09:01:53 WARN  app[][o.s.a.p.AbstractManagedProcess] Process exited with exit value [Web Server]: 134
2022.10.18 09:01:53 INFO  app[][o.s.a.SchedulerImpl] Process[Web Server] is stopped
2022.10.18 09:01:53 WARN  app[][o.s.a.p.AbstractManagedProcess] Process exited with exit value [ElasticSearch]: 143
2022.10.18 09:01:53 INFO  app[][o.s.a.SchedulerImpl] Process[ElasticSearch] is stopped
2022.10.18 09:01:53 INFO  app[][o.s.a.SchedulerImpl] SonarQube is stopped

**Software Versions**
 - SonarQube Version: v9.7.0
 - Plugin Version: v1.12.0

**Additional context**

**Comment by pablosguajardo** [comment]
It does not work for me either:
18:16:54.109 [main] INFO com.github.mc1arke.sonarqube.plugin.CommunityBranchAgent - Loading agent
Exception in thread "main" java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:513)
        at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:525)
Caused by: java.lang.ClassNotFoundException: org.sonar.server.almsettings.MultipleAlmFeatureProvider
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:315)
        at com.github.mc1arke.sonarqube.plugin.CommunityBranchAgent.redefineEdition(CommunityBranchAgent.java:93)
        at com.github.mc1arke.sonarqube.plugin.CommunityBranchAgent.premain(CommunityBranchAgent.java:56)
        ... 6 more
*** java.lang.instrument ASSERTION FAILED ***: "result" with message agent load/premain call failed at ./src/java.instru
ment/share/native/libinstrument/JPLISAgent.c line: 422
FATAL ERROR in native method: processing of -javaagent failed, processJavaStart failed

**Comment by mc1arke** [comment]
https://github.com/SonarSource/sonarqube/commit/57a8a41fa35a7949e9cb3582b60e368bd5253e2d has moved `MultipleAlmFeatureProvider` to `MultipleAlmFeature` so that will need an update to the Java Agent to use the new class and method name. I've not analysed the other changes in Sonarqube in this last release to see if there are any other breaking changes yet.

**Comment by pablosguajardo** [comment]
Thanks for answering. what should we do? do you have to create a new version of the plugin?

**Comment by 4n4n4s** [comment]
> Thanks for answering. what should we do? do you have to create a new version of the plugin?

Yes for now stick to 9.6 and don't update until there is a new version of the plugin available.

Example for docker you can use this tag [9.6-community](https://hub.docker.com/layers/library/sonarqube/9.6-community/images/sha256-2c14cb8b69da559bef28fdd1bcd022ecf59961fb30ef187ecc7f03d9f0e642b9?context=explore)

**Comment by scholz-afis** [comment]
Maybe you should highlight the 9.7 issue in der `Compatibility`section of https://github.com/mc1arke/sonarqube-community-branch-plugin/blob/master/README.md ?

**Comment by lorirava** [comment]
> > Thanks for answering. what should we do? do you have to create a new version of the plugin?
> 
> Yes for now stick to 9.6 and don't update until there is a new version of the plugin available.
> 
> Example for docker you can use this tag [9.6-community](https://hub.docker.com/layers/library/sonarqube/9.6-community/images/sha256-2c14cb8b69da559bef28fdd1bcd022ecf59961fb30ef187ecc7f03d9f0e642b9?context=explore)

if you use kubernetes helm chart, is it possible to use 9.6 version? I got a server error that say me to use the latest version...

**Comment by scholz-afis** [comment]
@lorirava you might find these commands helpful:
```
helm search repo sonarqube -l
helm upgrade --install sonarqube sonarqube/sonarqube --namespace sonarqube --version 5.0.6+370
```

**Comment by Mistic92** [comment]
I can't downgrade from 9.7. Do we have any ETA for fix?

**Comment by mgexm** [comment]
I was completely lucky. We backup the filesystem that is used so we restored. However to you point @Mistic92 I was getting the exact "cannot downgrade a node from version [7.17.5] to version [7.17.4]" which was caused by an attempt to upgrade to SQ 9.7.0

2022.10.25 16:58:09 INFO  es[][o.e.e.NodeEnvironment] using [1] data paths, mounts [[/opt/sonarqube/data (/dev/xvdb)]], net usable_space [91.7gb], net total_space [97.8gb], types [ext4]
2022.10.25 16:58:09 INFO  es[][o.e.e.NodeEnvironment] heap size [512mb], compressed ordinary object pointers [true]
2022.10.25 16:58:09 ERROR es[][o.e.b.ElasticsearchUncaughtExceptionHandler] uncaught exception in thread [main]
org.elasticsearch.bootstrap.StartupException: java.lang.IllegalStateException: cannot downgrade a node from version [7.17.5] to version [7.17.4]
	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:170) ~[elasticsearch-7.17.4.jar:7.17.4]
	at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:157) ~[elasticsearch-7.17.4.jar:7.17.4]
	at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:77) ~[elasticsearch-7.17.4.jar:7.17.4]
	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:112) ~[elasticsearch-cli-7.17.4.jar:7.17.4]
	at org.elasticsearch.cli.Command.main(Command.java:77) ~[elasticsearch-cli-7.17.4.jar:7.17.4]
	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:122) ~[elasticsearch-7.17.4.jar:7.17.4]
	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:80) ~[elasticsearch-7.17.4.jar:7.17.4]
Caused by: java.lang.IllegalStateException: cannot downgrade a node from version [7.17.5] to version [7.17.4]
	at org.elasticsearch.env.NodeMetadata.upgradeToCurrentVersion(NodeMetadata.java:95) ~[elasticsearch-7.17.4.jar:7.17.4]
	at org.elasticsearch.env.NodeEnvironment.loadNodeMetadata(NodeEnvironment.java:484) ~[elasticsearch-7.17.4.jar:7.17.4]
	at org.elasticsearch.env.NodeEnvironment.<init>(NodeEnvironment.java:356) ~[elasticsearch-7.17.4.jar:7.17.4]
	at org.elasticsearch.node.Node.<init>(Node.java:429) ~[elasticsearch-7.17.4.jar:7.17.4]
	at org.elasticsearch.node.Node.<init>(Node.java:309) ~[elasticsearch-7.17.4.jar:7.17.4]
	at org.elasticsearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:234) ~[elasticsearch-7.17.4.jar:7.17.4]
	at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:234) ~[elasticsearch-7.17.4.jar:7.17.4]
	at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:434) ~[elasticsearch-7.17.4.jar:7.17.4]
	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:166) ~[elasticsearch-7.17.4.jar:7.17.4]
	... 6 more
uncaught exception in thread [main]
java.lang.IllegalStateException: cannot downgrade a node from version [7.17.5] to version [7.17.4]
	at org.elasticsearch.env.NodeMetadata.upgradeToCurrentVersion(NodeMetadata.java:95)
	at org.elasticsearch.env.NodeEnvironment.loadNodeMetadata(NodeEnvironment.java:484)
	at org.elasticsearch.env.NodeEnvironment.<init>(NodeEnvironment.java:356)
	at org.elasticsearch.node.Node.<init>(Node.java:429)
	at org.elasticsearch.node.Node.<init>(Node.java:309)
	at org.elasticsearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:234)
	at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:234)
	at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:434)
	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:166)
	at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:157)
	at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:77)
	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:112)
	at org.elasticsearch.cli.Command.main(Command.java:77)
	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:122)
	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:80)
For complete error details, refer to the log at /opt/sonarqube/logs/sonarqube.log

**Comment by pablosguajardo** [comment]
I have sonar 9.7 installed with mssql, can I downgrade from version 9.7 to version 9.6? should I delete the base and start from scratch?

**Comment by githubcdr** [comment]
The downgrade issue does not seem to be related to the Postgres database, we restored a previous version and still receive this error.

**Comment by githubcdr** [comment]
I got it working by removing the Elasticsearch data created by Sonarqube, after this it will start with v9.6.

**Comment by pablosguajardo** [comment]
> I got it working by removing the Elasticsearch data created by Sonarqube, after this it will start with v9.6.

thanks for your answer. How did you removing this data?

**Comment by thibaut-trouve** [comment]
If you dont care about your current data you can just drop your sonarqube database and recreate it with the right privileges (as mentioned in the docs).

Example on Postgresql: 
```sql
CREATE USER sonar WITH ENCRYPTED PASSWORD '************'; -- Skip if you already have a user.
DROP DATABASE sonarqube;
CREATE database sonarqube;
GRANT ALL PRIVILEGES ON DATABASE sonarqube TO sonar;
```

Warning: Do it only if you are ready to loose your current sonarqube data!

If you're running Sonarqube using Docker and want a full reset, you can in addition, delete your volumes
`docker-compose down -v`

**Comment by aminoacidity** [comment]
> > thanks for your answer. How did you removing this data?
> 
> If you dont care about your current data you can just drop your sonarqube database and recreate it with the right privileges (as mentioned in the docs).
> 
> Example on Postgresql:
> 
> ```sql
> CREATE USER sonar WITH ENCRYPTED PASSWORD '************'; -- Skip if you already have a user.
> DROP DATABASE sonarqube;
> CREATE database sonarqube;
> GRANT ALL PRIVILEGES ON DATABASE sonarqube TO sonar;
> ```
> 
> Warning: Do it only if you are ready to loose your current sonarqube data!
> 
> If you're running Sonarqube using Docker and want a full reset, you can in addition, delete your volumes `docker-compose down -v`

Hi, I was referring to elasticsearch data cleanup that was referred to by the poster. I am guessing deleting the es folder should suffice.

**Comment by aminoacidity** [comment]
After reverting to the 9.6 version, deleted the ES data which in my case was mapped to /var/sonarqube/data/es7. Sonarqube started just fine.

**Comment by Mistic92** [comment]
Hi, any ETA for that for people who can't delete history?:)

**Comment by mgexm** [comment]
9.7.1 is out ... does the plugin work with 9.7.1 ?

**Comment by MCMicS** [comment]
Don't think so.
It should be fixed on plugin side i guess

**Comment by jeremytbrun** [comment]
@mc1arke any ETA on an update to the plugin to be compatible with SonarQube 9.7.x?

**Comment by henrypham299** [comment]
@mc1arke do we have any update for this issue?

**Comment by f33r0** [comment]
9.7, 9.7.0, 9.7.1 - plugin doesn't work

**Comment by EdmundMorris** [comment]
Does anyone know of a patched fork while we wait for the fix here?

**Comment by marcelloceschia** [comment]
If someone helps me to test it later, I can try to do a working version for next week

**Comment by AdamMcIntosh** [comment]
@marcelloceschia I can help test

**Comment by Mistic92** [comment]
@marcelloceschia I can help too but will need to have fork to review code before adding to sonar.

**Comment by SETDEVELOPERS** [comment]
Hi @mc1arke, is this change the solution? https://github.com/mc1arke/sonarqube-community-branch-plugin/compare/master...jepp-igus:sonarqube-community-branch-plugin:master

**Comment by jeremytbrun** [comment]
It appears @mc1arke hasn't been active in this repo since June and only active twice in any of his other repos since then. That doesn't seem like a good sign...

**Comment by Mistic92** [comment]
After trying few forks [[1](https://github.com/BrandonCate/sonarqube-community-branch-plugin/commit/b3ad68388fac8bd72a7e5369d1b249b76b46eff6)] which code was building I'm getting error in logs 
```
java.lang.IllegalStateException: Current edition does not support branch feature
 at com.google.common.base.Preconditions.checkState(Preconditions.java:502)
 at org.sonar.server.ce.queue.BranchSupport.createComponentKey(BranchSupport.java:52)
 at org.sonar.server.ce.queue.ReportSubmitter.submit(ReportSubmitter.java:81)
 ```

**Comment by mathijs-dumon** [comment]
Just hit this issue as well - great to see this is already fixed!
Is there a timeline for the next release?

**Comment by dgokcin** [comment]
@AHagemannCK I just hit this issue and trying with the plugin you suggested, Although sonarqube boots succesfully, I can not change branches from the project menu. Can you browse different branches thropugh the UI? 

![image](https://user-images.githubusercontent.com/33603535/208390247-1138d829-2e79-4426-a15d-700e87679711.png)

**Comment by EdmundMorris** [comment]
> > Just hit this issue as well - great to see this is already fixed! Is there a timeline for the next release?
> 
> You can just download the snapshot: https://github.com/mc1arke/sonarqube-community-branch-plugin/actions/runs/3669639838 and install it manually.

The snapshot is working on my v9.7.1 build. Many thanks to those involved.

**Comment by dgokcin** [comment]
@AHagemannCK I figured it out. I am hosting sonarqube on AWS ECS Fargate. During my version update from 8.9 LTS to 9.7, I forgot about the required java additional options. Once I enabled them back, it worked like a charm.

Thank you everyone!

**Comment by OneCyrus** [comment]
did anyone try 9.8 already? it just released today

**Comment by Elanis** [comment]
Just upgraded to 9.7.1 (62043), and using the latest snapshot from master branch. 
I get a "Current edition does not support branch feature" in background tasks :/

(Scanning using dotnet-sonarscanner, if that's important)
Am I the only one ?

**Comment by LocalHeroPro** [comment]
> did anyone try 9.8 already? it just released today

For me, SQ v9.8 don't start with v1.12 plugin.

**Comment by psini** [comment]
When it will be released ?

**Comment by dgokcin** [comment]
> Just upgraded to 9.7.1 (62043), and using the latest snapshot from master branch. I get a "Current edition does not support branch feature" in background tasks :/
> 
> I can browse through branches, and removing data/es7 like Mistic92 didn't fix things.
> 
> (Scanning using dotnet-sonarscanner, if that's important) Am I the only one ?

You should not use the latest snapshot from the master branch.

> Just upgraded to 9.7.1 (62043), and using the latest snapshot from master branch. I get a "Current edition does not support branch feature" in background tasks :/
> 
> I can browse through branches, and removing data/es7 like Mistic92 didn't fix things.
> 
> (Scanning using dotnet-sonarscanner, if that's important) Am I the only one ?

I use the same sonarqube version with dotnet sonar scanner. Did you also update the versions of the plugin in the sonarqube run command? Adding my config for reference. Note that I renamed the snapshot as 1.13.0.jar and put it inside my docker container at the specified paths in the README of this repo.
```hcl
  app_command = ["-Dsonar.search.javaAdditionalOpts=-Dnode.store.allow_mmap=false",
    "-Dsonar.web.javaAdditionalOpts=-javaagent:./extensions/plugins/sonarqube-community-branch-plugin-1.13.0.jar=web",
    "-Dsonar.ce.javaAdditionalOpts=-javaagent:./extensions/plugins/sonarqube-community-branch-plugin-1.13.0.jar=ce"
  ]


```

**Comment by Elanis** [comment]
> > Just upgraded to 9.7.1 (62043), and using the latest snapshot from master branch. I get a "Current edition does not support branch feature" in background tasks :/
> > I can browse through branches, and removing data/es7 like Mistic92 didn't fix things.
> > (Scanning using dotnet-sonarscanner, if that's important) Am I the only one ?
> 
> I use the same sonarqube version with dotnet sonar scanner. Did you also update the versions of the plugin in the sonarqube run command? Adding my config for reference. Note that I renamed the snapshot as 1.13.0.jar and put it inside my docker container at the specified paths in the README of this repo.
> 
> ```terraform
>   app_command = ["-Dsonar.search.javaAdditionalOpts=-Dnode.store.allow_mmap=false",
>     "-Dsonar.web.javaAdditionalOpts=-javaagent:./extensions/plugins/sonarqube-community-branch-plugin-1.13.0.jar=web",
>     "-Dsonar.ce.javaAdditionalOpts=-javaagent:./extensions/plugins/sonarqube-community-branch-plugin-1.13.0.jar=ce"
>   ]
> ```

Seeing your config made me instantly understand my mistake, I forgot to add additionalOpts for CE (I only did for web) when I did migrate to 9.7.1, my bad. It's working now, thanks !

**Comment by AmitBaranes** [comment]
can you publish this version so we'll not need to build it manually?

**Comment by scholz-afis** [comment]
> can you publish this version so we'll not need to build it manually?

... and update the `README.md` regarding the current situation. Newcomers shouldn't get trapped with a broken 9.7+ installation in the first place.

**Comment by MI-LA01** [comment]
> can you publish this version so we'll not need to build it manually?

Any updates ?

**Comment by nixel2007** [comment]
You can grab snapshot from github actions build

**Comment by MI-LA01** [comment]
Hey @nixel2007, 
> You can grab snapshot from github actions build

Any reason to not get release and only -snap version ?

**Comment by nixel2007** [comment]
> Any reason to not get release and only -snap version ? 

Real life maybe? :)

Let's wait for author's response

**Comment by mc1arke** [comment]
As suggested, it's the balance of time between the work, life, and open source. I'll get a release out sometime soon, but my priority has to be the things that provide my income.

**Comment by mc1arke** [comment]
This has been released as part of 1.13.0
